sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoLifecycleTest.java

```diff
@@ -387,43 +367,84 @@ private ExceptionThrowingFn(MethodForException toThrow) {
       this.toThrow = toThrow;
     }
 
+    private static void validate(CallState... requiredCallStates) {
+      Map<Integer, DelayedCallStateTracker> callStates;
+      synchronized (ExceptionThrowingFn.class) {
```

Comment:
Started seeing failures like:
```
java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1469)
	at java.util.HashMap$ValueIterator.next(HashMap.java:1498)
	at java.util.AbstractCollection.toString(AbstractCollection.java:461)
	at java.lang.String.valueOf(String.java:2994)
	at java.lang.StringBuilder.append(StringBuilder.java:131)
	at org.apache.beam.sdk.transforms.ParDoLifecycleTest$ExceptionThrowingFn.validate(ParDoLifecycleTest.java:392)
	at org.apache.beam.sdk.transforms.ParDoLifecycleTest$ExceptionThrowingFn.access$300(ParDoLifecycleTest.java:356)
	at org.apache.beam.sdk.transforms.ParDoLifecycleTest.testTeardownCalledAfterExceptionInProcessElementStateful(ParDoLifecycleTest.java:269)
```

Filed https://issues.apache.org/jira/projects/BEAM/issues/BEAM-14027

Reply:
this might be related to the previous PR fixing this being rolled back. I'm trying to resubmit that PR

Reply:
Ah, the bug is simpler. In assertThat we should be referencing callStates (the cloned map) instead of ExceptionThrowingFn.callStateMap.

Reply:
sent pr/16991 to fix this